File: javascript/lib/gherkin/io.js
var stream = require('stream')
var es = require('event-stream')
var Gherkin = require('../..');

var compiler = new Gherkin.Compiler();
var parser = new Gherkin.Parser(new Gherkin.AstBuilder());
parser.stopAtFirstError = false;

module.exports = function io(input, output, options) {
  input
    .pipe(es.split())
    .pipe(es.mapSync(function (line) {
      return line ? JSON.parse(line) : undefined
    }))
    .pipe(new stream.Transform({
      objectMode: true,
      transform: function (event, _, callback) {
        if (event.type === 'source') {
          var gherkinDocument = parser.parse(event.data);

          if (options.printSource)
            this.push(event)

          if (options.printAst)
            this.push(gherkinDocument);

          if (options.printPickles) {
            var pickles = compiler.compile(gherkinDocument, event.uri);
            pickles.map(this.push.bind(this));
          }
        } else {
          this.push(event)
        }
      }
    }))
    .pipe(es.mapSync(JSON.stringify))
    .pipe(es.join('\n'))
    .pipe(output)
}

File: ruby/lib/gherkin/io.rb
require 'json'
require 'gherkin/parser'
require 'gherkin/pickles/compiler'

module Gherkin
  class IO
    def initialize(print_source, print_ast, print_pickles)
      @print_source, @print_ast, @print_pickles = print_source, print_ast, print_pickles
      @parser = Gherkin::Parser.new
      @compiler = Gherkin::Pickles::Compiler.new
    end

    def process(input, output)
      input.each_line do |line|
        event = JSON.parse(line)
        if (event['type'] == 'source')
          uri = event['uri']
          source = event['data']
          begin
            gherkin_document = @parser.parse(source)

            if (@print_source)
              output.puts(line)
            end
            if (@print_ast)
              output.puts(gherkin_document.to_json)
            end
            if (@print_pickles)
              pickles = @compiler.compile(gherkin_document, uri)
              pickles.each do |pickle|
                output.puts(pickle.to_json)
              end
            end
          rescue Gherkin::CompositeParserException => e
            print_errors(output, e.errors, uri)
          rescue Gherkin::ParserError => e
            print_errors(output, [e], uri)
          end
        else
        end
      end
    end

    def print_errors(output, errors, uri)
      errors.each do |error|
        output.puts(JSON.generate({
          type: "attachment",
          source: {
            uri: uri,
            start: {
              line: error.location[:line],
              column: error.location[:column]
            }
          },
          data: error.message,
          media: {
            encoding: "utf-8",
            type: "text/vnd.cucumber.stacktrace+plain"
          }
        }))
      end
    end
  end
end
